src/lem/mod.rs

```diff
@@ -422,9 +598,327 @@ mod tests {
         })
         .unwrap();
 
-        let expr = Ptr::num(Fr::from_u64(42));
-        let mut store = Store::default();
-        let witnesses = lem.eval(expr, &mut store).unwrap();
-        constrain_test_helper(&lem, &mut store, &witnesses);
+        constrain_test_helper(&lem, &Ptr::num(Fr::from_u64(42)), NumSlots::default());
+    }
+
+    #[test]
+    fn test_hash2_slots_simple() {
+        let lem = lem!(expr_in env_in cont_in {
+            match_tag expr_in {
+                Num => {
+                    let expr_out: Cons = hash2(expr_in, expr_in);
```

Comment:
Am I correctly understanding that the `lem!` macro accounts for the 'declaredtype' and constrains the tag?

Reply:
If I understood your question then yes. The `: Cons` part can't be omitted.

Reply:
Important detail: the `lem!` macro itself doesn't do any constraining. It simply instantiates the `LEM`